### PR TITLE
Unify `push!`, `append!` and `vcat` implementation.

### PR DESCRIPTION
**Pull request by rapus95**
See https://github.com/JuliaData/DataFrames.jl/issues/2032

There's still a lot of work to be done. But for easier tracking and feedback I open it now.

### CODE REVIEW

**Comment by bkamins** [comment]
Thank you for working on it. As noted earlier I would recommend to write down the target API specification for `vcat`, `push!` and `append!` (similarly to what you did in #2032 but there you were commenting about internal design and here we should a specification of how functions are intended to work). Actually we might add a separate section for `vcat`, `push!` and `append!` to the documentation like we did for indexing.

**Comment by rapus95** [comment]
I don't intend to change the API currently, though, I want to make the internals more capable (as stated in #2032). If you suggest some API change to "unleash" the full capabilities to the public, I'm open for suggestions. Though, I might first get the implementation done (better than later finding it unsolvable to be implemented performant) and already having a new API.

**Comment by bkamins** [comment]
I have quickly went through the PR and it seems you are touching `vcat` only (which is OK), but you are changing the API of `vcat` in it (which is also fine in general, but you say that you do not do it so I am not fully sure what is the intention).

If you intend to change only the internals and do not touch the API then it is fine - I guess then the starting point for a review is that CI should pass the old tests.

**Comment by rapus95** [comment]
Ah indeed, though, for now, much will change. When `vcat` is done I will try to generalize the helper functions in order to expand their usage onto `push!` and `append!`. At the end API I'll return to previous API if wanted (for now they just show, what the function would be capable of)

**Comment by bkamins** [comment]
OK.
Please note that from my side it is easiest to review PRs that are either documented in detail (so we know what we want in the end - this is most often the case for large changes that cannot be accomplished in a single PR) or self-contained (i.e. providing some functionality end-to-end that can be checked - this is most often the case for small PRs).

**Comment by rapus95** [comment]
I'll try to split up the whole PR into small logically separated parts where each should be verifiable in a simple fashion and the whole verification comes by modus ponens :D

**Comment by rapus95** [comment]
I've found a single change (that might be breaking but plausible in my opinion):
0-row but non-0-width dataframes are taken into account when designing the new column to the extent that the actual vcat/append! etc fail if the type of a 0-length column cannot be converted into the inferred type.
Tl;Dr those functions can fail even when appending a 0-length dataframe. I find that plausible because supplying a 0-length dataframe wouldn't be useful at all if you didn't want to influence the new table. Most useful case would be to "prepare" a typestable (in regards to not changing column types) appending of more data. You could then prepare that by merging with a 0-row signature ahead of time. Then we can guarantuee the following append to succeed if the supplied types had been supplied before.

**Comment by bkamins** [comment]
I cannot reproduce what you are describing:
```
julia> using DataFrames

julia> df1 = DataFrame(a=["1", "2"])
2×1 DataFrame
│ Row │ a      │
│     │ String │
├─────┼────────┤
│ 1   │ 1      │
│ 2   │ 2      │

julia> df2 = DataFrame(a=Int[])
0×1 DataFrame


julia> vcat(df1, df2)
2×1 DataFrame
│ Row │ a   │
│     │ Any │
├─────┼─────┤
│ 1   │ 1   │
│ 2   │ 2   │

julia> append!(df1, df2)
2×1 DataFrame
│ Row │ a      │
│     │ String │
├─────┼────────┤
│ 1   │ 1      │
│ 2   │ 2      │
```
(and there is no error in either of the cases - the behavior is different as `vcat` does type promotion and `append!` does not, but this is a separate issue).

Can you please supply example codes with what you had in mind (as supposedly this is something different). Showing: current behavior and the proposed behavior? Thank you.

**Comment by rapus95** [comment]
I've resolved it. There shouldn't be any difference. 😀

**Comment by bkamins** [comment]
This should be closed when #2152 is merged. We can open a new PR for post 1.0 functionality.

**Comment by bkamins** [comment]
OK we have merged #2152. If something more would to be added please open a new PR.
